README.md
# isInHammingDistance

Returns true if the hamming distance between two strings or arrays of equal length is smaller than the given limit.

## Install

```
npm install is-in-hamming-distance --save
```

## Usage

```js
import isInHammingDistance from "is-in-hamming-distance";

console.log(isInHammingDistance(3, "karolin", "kathrin"));
// => true
console.log(isInHammingDistance(1, "karolin", "kathrin"));
// => false
```

## License

MIT

## Links

-   Benchmark: [BENCHMARK.txt](./BENCHMARK.txt)
-   Wikipedia: [Hamming Distance](https://en.wikipedia.org/wiki/Hamming_distance)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/guilhermehn/is-in-hamming-distance/compare/v1.0.0...v2.0.0) (2023-04-25)

### ⚠ BREAKING CHANGES

-   uses ESM syntax.

### Bug Fixes

-   update prerelease script ([61a09cf](https://github.com/guilhermehn/is-in-hamming-distance/commit/61a09cf16c029a3b13cc1107482d7271609d1e17))

-   drop support for older Node versions ([d8fbb79](https://github.com/guilhermehn/is-in-hamming-distance/commit/d8fbb79960a04d813cc0bc1bae5c7988d3255dbb))

## 1.0.0 (2020-04-26)


spec.js
const isInHammingDistance = require(".");

describe("isInHammingDistance", () => {
	it("should return true if the hamming distance between strings is smaller than the limit parameter", () => {
		// https://en.wikipedia.org/wiki/Hamming_distance#Examples
		expect(isInHammingDistance(0, "karolin", "karolin")).toBe(true);
		expect(isInHammingDistance(3, "karolin", "kathrin")).toBe(true);
		expect(isInHammingDistance(3, "karolin", "kerstin")).toBe(true);
		expect(isInHammingDistance(2, "1011101", "1001001")).toBe(true);
		expect(isInHammingDistance(3, "2173896", "2233796")).toBe(true);
		expect(isInHammingDistance(+Infinity, "a", "b")).toBe(true);
		expect(isInHammingDistance(1, [1, 2, 3], [1, 2, 4])).toBe(true);
	});

	it("should return false if the hamming distance is greater than the limit parameter", () => {
		// https://en.wikipedia.org/wiki/Hamming_distance#Examples
		expect(isInHammingDistance(2, "karolin", "kathrin")).toBe(false);
		expect(isInHammingDistance(2, "karolin", "kerstin")).toBe(false);
		expect(isInHammingDistance(1, "1011101", "1001001")).toBe(false);
		expect(isInHammingDistance(2, "2173896", "2233796")).toBe(false);
		expect(isInHammingDistance(0, "a", "b")).toBe(false);
		expect(isInHammingDistance(1, [1, 2, 3], [1, 4, 5])).toBe(false);
	});
});


index.js
const assert = require("assert").strict;

/**
 * Returns true if the hamming distance between two strings or arrays of equal length is smaller than the given limit
 * @param {Number} - The distance limit between both values. If the difference is greater than this limit, this function will return false.
 * @param {String} - Some string
 * @param {String} - Some string
 * @throws
 * @returns {Boolean}
 * @see https://en.wikipedia.org/wiki/Hamming_distance
 * @example
 * const isInHammingDistance = require('is-in-hamming-distance');
 *
 * console.log(isInHammingDistance(3, 'karolin', 'kathrin'));
 * // => true
 * console.log(isInHammingDistance(1, 'karolin', 'kathrin'));
 * // => false
 */
function isInHammingDistance(limit, a, b) {
	assert.strictEqual(a.length, b.length, "Values should have equal length");

	let distance = 0;

	for (let i = 0; i < a.length; i++) {
		if (a[i] !== b[i]) {
			if (++distance > limit) {
				return false;
			}
		}
	}
	return true;
}

module.exports = isInHammingDistance;


index.mjs
import { strict as assert } from "node:assert";

/**
 * Returns true if the hamming distance between two strings or arrays of equal length is smaller than the given limit
 * @param {Number} - The distance limit between both values. If the difference is greater than this limit, this function will return false.
 * @param {String} - Some string
 * @param {String} - Some string
 * @throws
 * @returns {Boolean}
 * @see https://en.wikipedia.org/wiki/Hamming_distance
 * @example
 * const isInHammingDistance = require('is-in-hamming-distance');
 *
 * console.log(isInHammingDistance(3, 'karolin', 'kathrin'));
 * // => true
 * console.log(isInHammingDistance(1, 'karolin', 'kathrin'));
 * // => false
 */
export default function isInHammingDistance(limit, a, b) {
    assert.strictEqual(a.length, b.length, "Values should have equal length");

    let distance = 0;

    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            if (++distance > limit) {
                return false;
            }
        }
    }
    return true;
}


spec.mjs
import isInHammingDistance from "./index.mjs";
import { describe, it } from "node:test";
import { strict as assert } from "node:assert";

describe("isInHammingDistance", () => {
	it("should return true if the hamming distance between strings is smaller than the limit parameter", () => {
		// https://en.wikipedia.org/wiki/Hamming_distance#Examples
		assert.ok(isInHammingDistance(0, "karolin", "karolin"));
		assert.ok(isInHammingDistance(3, "karolin", "kathrin"));
		assert.ok(isInHammingDistance(3, "karolin", "kerstin"));
		assert.ok(isInHammingDistance(2, "1011101", "1001001"));
		assert.ok(isInHammingDistance(3, "2173896", "2233796"));
		assert.ok(isInHammingDistance(+Infinity, "a", "b"));
		assert.ok(isInHammingDistance(1, [1, 2, 3], [1, 2, 4]));
	});

	it("should return false if the hamming distance is greater than the limit parameter", () => {
		// https://en.wikipedia.org/wiki/Hamming_distance#Examples
		assert.equal(isInHammingDistance(2, "karolin", "kathrin"), false);
		assert.equal(isInHammingDistance(2, "karolin", "kerstin"), false);
		assert.equal(isInHammingDistance(1, "1011101", "1001001"), false);
		assert.equal(isInHammingDistance(2, "2173896", "2233796"), false);
		assert.equal(isInHammingDistance(0, "a", "b"), false);
		assert.equal(isInHammingDistance(1, [1, 2, 3], [1, 4, 5]), false);
	});
});
